Explainer of Objects (#130)
tc39/proposal-temporal: explainer.md

# Temporal

## Timeline Objects

There is a subset of temporal objects that are tied to the absolute timeline. They represent a specific point in time.

> These are `Instant`, `OffsetDateTime` and `ZonedDateTime` in order of amount of detail information available in them.

Reviewer: ```suggestion
These are `Instant`, `OffsetDateTime`, and `ZonedDateTime` in order of amount of detail information available in them.
```

> Instant just represents an absolute point in time. No timezone or offset information is present. As such Intants have no concept of days or months or even hours.

Reviewer: ```suggestion
Instant just represents an absolute point in time. No timezone or offset information is present. As such Instants have no concept of days or months or even hours.
```

OffsetDateTime is the combination of an Instant with an offset from UTC. As such it has the ability to know about days, months, etc. However it does not know what timezone or locality it is in. As such it cannot know which daylight saving rules apply.

ZonedDateTime represents an OffsetDateTime combined with an *IANA Timezone*. With that added information ZonedDateTime is able to observe daylight savings rules.

### Instant

> An `Instant` is an object that specifies a specific point in time. For convenience of interoperatbility it uses *nanoseconds since the unix-epoch* to do so.

Reviewer: ```suggestion
An `Instant` is an object that specifies a specific point in time. For convenience of interoperability it uses *nanoseconds since the unix-epoch* to do so.
```

#### Instant() - constuctor

#### Instant.prototype.epochSeconds : number

> The `epochSeconds` property of an `Instant` object is readonly and represents the *seconds

Reviewer: when you say "readonly", do you mean it's a getter without a setter?

Author: I'll mke it clearer that all non-function properties are getters (i.e. readonly). I think a paragraph toward the beginning making clear that all temporal objects are immutable should do it.

since unix-epoch*.

#### Instant.prototype.epochMilliseconds: number

The `epochMilliseconds` property of an `Instant` object is readonly and represents the *milliseconds
since unix-epoch*.

#### Instant.prototype.epochMicroseconds: BigInt

The `epochMicroseconds` property of an `Instant` object is readonly and represents the *microseconds
since unix-epoch*.

#### Instant.prototype.epochNanoseconds: BigInt

The `epochNanoseconds` property of an `Instant` object is readonly and represents the *nanoseconds
since unix-epoch*.

> #### Instant.prototype.witZone(zone: string) : ZonedDateTime

Reviewer: ```suggestion
#### Instant.prototype.withZone(zone: string) : ZonedDateTime
```

#### Instant.fromString(iso: string) : Instant

Parses a `string` that must be in the same ISO-8601 format as produced by `Instant.prototype.toString()` and creates a new `Instant` object from it.

#### Instant.fromEpochSeconds(epochseconds: number) : Instant

Equivalent to `Instant.fromMilliseconds(seconds * 1000)`.

#### Instant.fromEpochMilliseconds(epochmillis: number) : Instant

Equivalent to `Instant.fromEpochMicroseconds(BigInt(epochmillis) * 1000n)`.

#### Instant.fromEpochMicroseconds(epochmicros: BigInt) : Instant

Equivalent to `Instant.fromEpochNanoseconds(epochmicros * 1000n)`.

#### Instant.fromEpochNanoseconds(epochnanos: BigInt) : Instant

Equivalent to `new Instant(epochnanos)`.

### OffsetDateTime

An `OffsetDateTime` is an object that specifies a specific point in time with a specific offset from *UTC*.
It bases this on an `Instant` and an offsett `string`.

#### OffsetDateTime(instant: Instant, offset: string) - constructor

The constructor may only be called as such. It takes two arguments. The first is an instance
of `Instant` and the second is a `string` representing a valid time offset.

#### OffsetDateTime.prototype.instant : BigInt

> The instance of `Instant` that was used to create this `OffsetDateTime`.

Reviewer: is this also readonly, ie, an accessor?

Author: Yes

#### OffsetDateTime.prototype.offset : string

The time-zone offset represented as a `string`.

This must always have the format: `sign``hours`:`minutes` where

- `sign` is eitner `+` or `-`
- `hours` is the hours offset 0-padded to 2 digits.
- `minutes` is the minutes offset 0-padded to 2 digits.

Examples: `+00:00`, `-04:00`, `+03:00`, ...

#### OffsetDateTime.prototype.year: number

> The `.year` property represents the year of the `OffsetDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: if it's necessary to specify "readonly" above, should it be present on all these as well?

Author: Moved to general section

#### OffsetDateTime.prototype.nanosecond: number

The `.nanosecond` property represents the nanoseconds of the microseconds of the `OffsetDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

#### OffsetDateTime.prototype.dayOfWeek: number

The `.dayOfWeek` property represents the day of the week according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day where Monday is `1` and `Sunday` is `7` in accordance with ISO-8601. 

#### OffsetDateTime.prototype.dayOfYear: number

The `.dayOfYear` property represents the ordinal day of the Gregorian year according to ISO-8601.

#### OffsetDateTime.prototype.weekOfYear: number

The `.weekOfYear` property represents the ISO week-number. Beware that dates at the begining of a year may be part of a week from the preceding year, and dates at the end of a year may be part of a week at the beginning of the next year, as the first week of any year is defined as the week that contains the first Thursday of the week.

> #### OffsetDateTime.prototype.plus(data: IntervalLike): ZonedDateTime

Reviewer: can we link all these inline (IntervalLike, DateLike, etc) to their respective sections?

Author: 👍

Creates a new `ZonedDateTime` object by adding (subtracting for negative values) values to its members. The specified values must be numeric if specified.

The algorithm is such that:

1. the individual values are added to the existing values.
2. the range of `nanoseconds` is ensured to be between 0 and 999 by adjusting the `microseconds`
3. the range of `microseconds` is ensured to be between 0 and 999 by adjusting the `milliseconds`
4. the range of `milliseconds` is ensured to be between 0 and 999 by adjusting the `seconds`
5. the range of `seconds` is ensured to be between 0 and 59 by adjusting the `minutes`
6. the range of `minutes` is ensured to be between 0 and 59 by adjusting the `hours`
7. the range of `hours` is ensured to be between 0 and 23 by adjusting `days`
8. the range of `days` is ensured to be between 1 and 29-31 depending on the month by adjusting `month`
9. the range of `months` is ensured to be between 1 and 12 by adjusting the `years`.

#### OffsetDateTime.prototype.minus(other: ZonedDateTime): Interval

> #### OffsetDateTime.prototype.withZone(iana: string) : ZonedDateTime

Reviewer: does this throw when `iana` is invalid? is there a non-throwing method of validating that i have a string in the correct format?

Author: Yes & Yes : `ZonedDateTime.isValidTimezone()`

#### OffsetDateTime.prototype.toString() : string

> This creates an ISO-8601 string in the following format **`year`-`month`-`day`T`hours`:`minutes`:`seconds`.`nanoseconds``offset`**.

Reviewer: would this format be more intuitive to JS devs?
```suggestion
This creates an ISO-8601 string in the following format:
```js
${year}-${month}-${day}T${hours}:${minutes}:${seconds}.${nanoseconds}${offset}
```
```
(the suggestion might not display correctly due to the backticks)

> The `year` is 0-padded to a minimum of 4 digits. `month`, `day`, `hours`, `minutes`, `seconds` are 0-padded to a minimum of 2 digits. `nanoseconds` is 0-padded to a minimum of 9 digits. The `offset` is the timezone offset as created by `zoned.offset.

Reviewer: ```suggestion
The `year` is 0-padded to a minimum of 4 digits. `month`, `day`, `hours`, `minutes`, `seconds` are 0-padded to a minimum of 2 digits. `nanoseconds` is 0-padded to a minimum of 9 digits. The `offset` is the timezone offset as created by `zoned.offset`.
```
this will fix the next few lines of formatting issues

Examples:

- `1976-11-18T15:23:30.123456789+01:00` - created with `+01:00` offset

#### OffsetDateTime.prototype.toJSON() : string

Equivalent to `OffsetDateTime.prototype.toString() : string`

#### OffsetDateTime.fromString(iso: string) : ZonedDateTime

Parses a `string` in the specific ISO-8601 format emitted by `OffsetDateTime.prototype.toString()`.

#### OffsetDateTime.fromZonedDateTime(zoned : ZonedDateTime) : OffsetDateTime

### ZonedDateTime

> An `ZonedDateTime` is an object that specifies a specific point in time with an *IANA timezone*.

Reviewer: ```suggestion
A `ZonedDateTime` is an object that specifies a specific point in time with an *IANA timezone*.
```

It bases this on an `Instant` and an iana `string`.

#### ZonedDateTime(instant: Instant, zone: string) - constructor

The constructor may only be called as such. It takes two arguments. The first is an instance
of `Instant` and the second is a `string` representing a valid IANA timezone.

#### ZonedDateTime.prototype.instant : BigInt

The instance of `Instant` that was used to create this `ZonedDateTime`.

#### ZonedDateTime.prototype.timeZone : string
#### ZonedDateTime.prototype.offset : string

The time-zone offset represented as a `string`.

This must always have the format: `sign``hours`:`minutes` where

> - `sign` is eitner `+` or `-`

Reviewer: ```suggestion
- `sign` is either `+` or `-`
```

- `hours` is the hours offset 0-padded to 2 digits.
- `minutes` is the minutes offset 0-padded to 2 digits.

Examples: `+00:00`, `-04:00`, `+03:00`, ...

#### ZonedDateTime.prototype.year: number

The `.year` property represents the year of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

#### ZonedDateTime.prototype.month: number

> The `.month` property represents the month of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.month` property represents the month of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.day: number

> The `.day` property represents the day of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.day` property represents the day of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.hour: number

> The `.hour` property represents the hour of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.hour` property represents the hour of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.minute: number

> The `.minute` property represents the minute of the hour of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.minute` property represents the minute of the hour of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.second: number

> The `.second` property represents the seconds of the minute of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.second` property represents the seconds of the minute of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.millisecond: number

> The `.millisecond` property represents the milliseconds of the seconds of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.millisecond` property represents the milliseconds of the seconds of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.microsecond: number

> The `.microsecond` property represents the microseconds of the milliseconds of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.microsecond` property represents the microseconds of the milliseconds of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

#### ZonedDateTime.prototype.nanosecond: number

> The `.nanosecond` property represents the nanoseconds of the microseconds of the `ZonedDateTime `  according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.

Reviewer: ```suggestion
The `.nanosecond` property represents the nanoseconds of the microseconds of the `ZonedDateTime` according to the proleptic Gregorian calendar with a midnight to midnight 24 hour day.
```

calculated by `.dayOfWeek`. The `year` may be one year before/after the `.year` property of the
`CivilDate` if the specified date is part of the last week of the previous year or the first
week of the following year.

#### CivilDate.prototype.toOrdinalDateString() : string

`.toOrdinalDateString()` creates an ISO-8601 compliant strung in the format:
**`year`-`day-of-year`**.

The `year` is 0-padded to a minimum of 4 digits. `dof-of-year` is 0-padded to a minimum of 3 digits.

The `day-of-year` is the ordinal day as calculated by `.dayOfYear`.

#### CivilDate.fromDateString(isostring : string): string

Creates a new `CivilDate` by parsing an ISO-8601 string in the format created by `.toDateString()`.

#### CivilDate.fromWeekDateString(isostring : string): string

Creates a new `CivilDate` by parsing an ISO-8601 string in the format created by `.toWeekDateString()`.

#### CivilDate.fromOrdinalDateString(isostring : string): string

Creates a new `CivilDate` by parsing an ISO-8601 string in the format created by `.toOrdinalDateString()`.

#### CivilDate.fromString(isostring: string): CivilDate

> Creates a new `CivilDate` by parsing an ISO-8601 string in the one of the formats created `.toDateString()`, `.toWeekDateString()` or `.toOrdinalDateString()`.

Reviewer: ```suggestion
Creates a new `CivilDate` by parsing an ISO-8601 string in the one of the formats created `.toDateString()`, `.toWeekDateString()`, or `.toOrdinalDateString()`.
```

### CivilTime

`CivilTime` (and its siblings) represents a time corresponding to the requirements of ISO-8601.

This means specifically that days are 24 hours long and begin and end at midnight.

#### CivilTime() - constructor

The constructor may only be called as such. It takes between 2 and 4 numeric arguments.

- `hours` hour of the day
- `minutes` minutes of the hour
- `seconds` seconds of the minutes *(default: 0)*
- `milliseconds` milliseconds of the second *(default: 0)*
- `microseconds` microseconds of the millisecond *(default: 0)*

> - `nanoseconds` nanoseconds of the microseconds *(default: 0)

Reviewer: ```suggestion
- `nanoseconds` nanoseconds of the microseconds *(default: 0)*
```
this should fix many of the following formatting issues

The specified values must be numeric if specified.

The algorithm is such that:
 1. the individual values are added to the existing values.
 2. the range of `nanoseconds` is ensured to be between 0 and 999 by adjusting the `microseconds`
 3. the range of `microseconds` is ensured to be between 0 and 999 by adjusting the `milliseconds`
 4. the range of `milliseconds` is ensured to be between 0 and 999 by adjusting the `seconds`
 5. the range of `seconds` is ensured to be between 0 and 59 by adjusting the `minutes`
 6. the range of `minutes` is ensured to be between 0 and 59 by adjusting the `hours`
 7. the range of `hours` is ensured to be between 0 and 23

#### CivilTime.prototype.with(values: TimeLike) : CivilTime

Creates a new `CivilTime` object by overriding specified values to its members.
The specified values must be numeric if specified.

#### CivilTime.prototype.withDate(date : CivilDate) : CivilDateTime

Combines this `CivilTime` with the passed `CivilDate` to create a new `CivilDateTime` object.

#### CivilTime.prototype.toString() : string

`.toString()` creates an ISO-8601 compliant string in the format:
**`hour`:`minute`:`second`.`nanosecond`**.

> The `hours`, `minutes` and `seconds` are 0-padded to a minimum of 2 digits.

Reviewer: ```suggestion
The `hours`, `minutes`, and `seconds` are 0-padded to a minimum of 2 digits.
```

`nanoseconds` is 0-padded to a minimum of 9 digits.

#### CivilTime.prototype.toJSON() : string

Equivalent to `datetime.toString()`

#### CivilTime.fromString(isostring : string): string

Creates a new `CivilTime` by parsing an ISO-8601 string in the format created by `.toString()`.

### CivilDateTime

`CivilDateTime` (and its siblings) represents a date and time corresponding to the requirements of ISO-8601.

This means specifically that it uses the *proleptic Gregorian calendar* whose days are 24 hours
long and begin and end at midnight.

#### CivilDateTime() - constructor

The constructor may only be called as such. It takes between 5 and 7 numeric arguments.

- `year` the Gregorian year
- `month` the Gregorian month
- `day` the Gregorian day of the month
- `hours` hour of the day
- `minutes` minutes of the hour
- `seconds` seconds of the minutes *(default: 0)*
- `milliseconds` milliseconds of the seconds *(default: 0)*
- `microseconds` microseconds of the milliseconds *(default: 0)*

> - `nanoseconds` nanoseconds of the microseconds *(default: 0)

Reviewer: ```suggestion
- `nanoseconds` nanoseconds of the microseconds *(default: 0)*
```

# Temporal

Temporal is the proposal of a new date & time handling API for ECMA-Script. The design principles are:

- All temporal APIs are non-mutating.  All temporal objects are effectively immutable.
- All date values are based on the [Proleptic Gregorian Calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar).  Other calendar systems are out-of-scope for this proposal.  However, we will consider how future APIs may interact with this one such that extending it to support other calendars may be possible in a future proposal.
- All time-of-day values are based on a standard 24-hour clock.
- [Leap seconds](https://en.wikipedia.org/wiki/Leap_second) are not represented.

## Timeline Objects

> There is a subset of temporal objects that are tied to the absolute timeline. They represent a specific point in time.

Reviewer 2: What exactly is the 'absolute timeline'? How does it differ from a relative timeline? Googling for ‘absolute timeline’ doesn’t give much info. Assuming people might apply common sense to figure about what we mean by 'absolute timeline' might not work at scale (especially for people whose native language is not english). So if at all possible, it might be good to provide a link explaining it?

Author: Point taken.

These are `Instant`, `OffsetDateTime`, and `ZonedDateTime` in order of amount of detail information available in them.

> Instant just represents an absolute point in time. No timezone or offset information is present. As such Instants have no concept of days or months or even hours.

Reviewer 2: ```suggestion
`Instant` just represents an absolute point in time. No timezone or offset information is present. As such Instants have no concept of days or months or even hours.
```

> OffsetDateTime is the combination of an Instant with an offset from UTC. As such it has the ability to know about days, months, etc. However it does not know what timezone or locality it is in. As such it cannot know which daylight saving rules apply.

Reviewer 2: ```suggestion
`OffsetDateTime` is the combination of an Instant with an offset from UTC. As such it has the ability to know about days, months, etc. However it does not know what timezone or locality it is in. As such it cannot know which daylight saving rules apply.
```

> ZonedDateTime represents an OffsetDateTime combined with an *IANA Timezone*. With that added information ZonedDateTime is able to observe daylight savings rules.

Reviewer 2: ```suggestion
`ZonedDateTime` represents an `OffsetDateTime` combined with an *[IANA](https://www.iana.org/time-zones) Timezone*. With that added information `ZonedDateTime` is able to observe daylight savings rules.
```

### Instant
An `Instant` is an object that specifies a specific point in time. For convenience of interoperability it uses *nanoseconds since the unix-epoch* to do so.
#### Instant() - constuctor

#### Instant.prototype.epochSeconds : number

The `epochSeconds` property of an `Instant` object is readonly and represents the *seconds
since unix-epoch*.

#### Instant.prototype.epochMilliseconds: number

The `epochMilliseconds` property of an `Instant` object is readonly and represents the *milliseconds
since unix-epoch*.

#### Instant.prototype.epochMicroseconds: BigInt

The `epochMicroseconds` property of an `Instant` object is readonly and represents the *microseconds
since unix-epoch*.

#### Instant.prototype.epochNanoseconds: BigInt

The `epochNanoseconds` property of an `Instant` object is readonly and represents the *nanoseconds

> since unix-epoch*.

Reviewer 3: I've never heard this called the POSIX epoch before...

> since unix-epoch*.

Reviewer 4: I don't feel extremely strongly about this, but do believe "POSIX epoch" to be more accurate. The term "epoch" is defined at http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_150 (I intended to link to that in a suggestion right under the definition of Instant here, but apparently missed it), inside the specification that self-identifies as "_POSIX.1-2017 is simultaneously IEEE Std 1003.1™-2017 and The Open Group Technical Standard Base Specifications, Issue 7_". The Single UNIX specification is tightly related to that in a way that is likely no longer separable (cf. http://www.unix.org/version4/overview.html and http://www.unix.org/version4/ieee_std.html ), but the latter is technically based upon it rather than being identical, making POSIX the more fundamental of the two. It also has the practical benefit of being linkable.

> since unix-epoch*.

Reviewer 3: Which do you think is the term that in more common use? The more common term might be more easily understandable in documentation.

> since unix-epoch*.

Reviewer 4: I think both terms are so far from common use that their relative frequency is irrelevant. Any first reference to an epoch instant should expand to its actual definition (e.g., as "midnight at the beginning of 01 January, 1970 UTC" in https://tc39.github.io/ecma262/#sec-time-values-and-time-range ).

> since unix-epoch*.

Reviewer 3: Defining the term in this documentation sounds like a good idea to me. Personally, I heard about the "Unix epoch" several years ago (maybe when I was in high school?), long before I learned about the details of timezones or the other things that went into the design of Temporal. This thread is the first time I heard of the "POSIX epoch".

#### Instant.prototype.withZone(zone: string) : ZonedDateTime

This creates a `ZonedDateTime` by applying a *iana timezone* to the instant.

This is equivalent to `new ZonedDateTime(instant, zone)`

> #### Instant.prototype.withOffset(offset: string) : OffsetDateTime

Reviewer 4: Would it be fair to say that you're shooting for a model in which `.with*` methods tend to expect specific scalar-valued arguments for up-conversion as opposed to `.with` methods accepting objects from which to read specifically-named properties for same-type creation? If so, perhaps we should define independent hour/minute/second/etc. parameters for `.withTime`.

#### Instant.fromEpochSeconds(epochseconds: number) : Instant

Equivalent to `Instant.fromMilliseconds(seconds * 1000)`.

#### Instant.fromEpochMilliseconds(epochmillis: number) : Instant

Equivalent to `Instant.fromEpochMicroseconds(BigInt(epochmillis) * 1000n)`.

#### Instant.fromEpochMicroseconds(epochmicros: BigInt) : Instant

Equivalent to `Instant.fromEpochNanoseconds(epochmicros * 1000n)`.

#### Instant.fromEpochNanoseconds(epochnanos: BigInt) : Instant

Equivalent to `new Instant(epochnanos)`.

### OffsetDateTime

An `OffsetDateTime` is an object that specifies a specific point in time with a specific offset from *UTC*.
It bases this on an `Instant` and an offsett `string`.

#### OffsetDateTime(instant: Instant, offset: string) - constructor

The constructor may only be called as such. It takes two arguments. The first is an instance
of `Instant` and the second is a `string` representing a valid time offset.

#### OffsetDateTime.prototype.instant : BigInt

The instance of `Instant` that was used to create this `OffsetDateTime`.

#### OffsetDateTime.prototype.offset : string

> The time-zone offset represented as a `string`.

Author: I was hoping to. They were created/useful only in misshapen timezone handling in `Date`. I consider them a bad relic. (Convince me otherwise 😄 )

> The time-zone offset represented as a `string`.

Reviewer 4: I don't know if I can convince you, but I imagine that _not_ having it will create cases where authors manually perform awkward calculations like `offsetMinutes = dateTime.offset.split(":").reduce((sum, part, i) => (sum || -0) + Math.sign(Infinity/sum) * part * 60**(1-i), 0)` or `offsetSeconds = CivilDateTime.from(dateTime).withTimeZone("Etc/UTC").instant.epochSeconds - dateTime.instant.epochSeconds`, although we might be able to mitigate that with `minus`.

#### OffsetDateTime.prototype.plus(data: IntervalLike): ZonedDateTime

Creates a new `ZonedDateTime` object by adding (subtracting for negative values) values to its members. The specified values must be numeric if specified.

The algorithm is such that:

1. the individual values are added to the existing values.
2. the range of `nanoseconds` is ensured to be between 0 and 999 by adjusting the `microseconds`
3. the range of `microseconds` is ensured to be between 0 and 999 by adjusting the `milliseconds`
4. the range of `milliseconds` is ensured to be between 0 and 999 by adjusting the `seconds`
5. the range of `seconds` is ensured to be between 0 and 59 by adjusting the `minutes`
6. the range of `minutes` is ensured to be between 0 and 59 by adjusting the `hours`
7. the range of `hours` is ensured to be between 0 and 23 by adjusting `days`
8. the range of `days` is ensured to be between 1 and 29-31 depending on the month by adjusting `month`

> 9. the range of `months` is ensured to be between 1 and 12 by adjusting the `years`.

Reviewer 4: The `plus` algorithm descriptions need rewording to reflect the differential naming of properties between ~~Interval~~ Duration vs. other types (plural in the former, singular in the latter).

#### OffsetDateTime.prototype.minus(other: ZonedDateTime): Interval
#### OffsetDateTime.prototype.withZone(iana: string) : ZonedDateTime
#### OffsetDateTime.prototype.toString() : string

This creates an ISO-8601 string in the following format:
```js

> ${year}-${month}-${day}T${hours}:${minutes}:${seconds}.${nanoseconds}${offset}

Reviewer 4: This line needs to be followed by <code>```</code> to terminate the code block.

The `year` is 0-padded to a minimum of 4 digits. `month`, `day`, `hours`, `minutes`, `seconds` are 0-padded to a minimum of 2 digits. `nanoseconds` is 0-padded to a minimum of 9 digits. The `offset` is the timezone offset as created by `zoned.offset`.

Examples:

- `1976-11-18T15:23:30.123456789+01:00` - created with `+01:00` offset

#### OffsetDateTime.prototype.toJSON() : string

Equivalent to `OffsetDateTime.prototype.toString() : string`

#### OffsetDateTime.fromString(iso: string) : ZonedDateTime

> Parses a `string` in the specific ISO-8601 format emitted by `OffsetDateTime.prototype.toString()`.

Reviewer 4: ```suggestion
Creates a new `OffsetDateTime` object from parsing a string that must be in the same ISO 8601 format used for `OffsetDateTime.prototype.toString()`, with precision at or exceeding the minute level and non-empty non-"Z" UTC offset representation.
```

#### OffsetDateTime.fromZonedDateTime(zoned : ZonedDateTime) : OffsetDateTime

### ZonedDateTime

An `ZonedDateTime` is an object that specifies a specific point in time with an *IANA timezone*.
It bases this on an `Instant` and an iana `string`.

#### ZonedDateTime(instant: Instant, zone: string) - constructor

The constructor may only be called as such. It takes two arguments. The first is an instance
of `Instant` and the second is a `string` representing a valid IANA timezone.

#### ZonedDateTime.prototype.instant : BigInt

> The instance of `Instant` that was used to create this `ZonedDateTime`.

Reviewer 4: Could we restructure this to eliminate the redundancy via text like "`ZonedDateTime.prototype` includes all properties and methods of `OffsetDateTime.prototype`, plus the following additions and replacements"?

The specified values must be numeric if specified.

The algorithm is such that:

1. the individual values are added to the existing values.
2. the range of `nanoseconds` is ensured to be between 0 and 999 by adjusting the `microseconds`
3. the range of `microseconds` is ensured to be between 0 and 999 by adjusting the `milliseconds`
4. the range of `milliseconds` is ensured to be between 0 and 999 by adjusting the `seconds`
5. the range of `seconds` is ensured to be between 0 and 59 by adjusting the `minutes`
6. the range of `minutes` is ensured to be between 0 and 59 by adjusting the `hours`
7. the range of `hours` is ensured to be between 0 and 23 by adjusting `days`
8. the range of `days` is ensured to be between 1 and 29-31 depending on the month by adjusting `month`
9. the range of `months` is ensured to be between 1 and 12 by adjusting the `years`.

#### ZonedDateTime.prototype.minus(other: ZonedDateTime): Interval
#### ZonedDateTime.prototype.toString() : string

This creates an ISO-8601 string in the following format **`year`-`month`-`day`T`hours`:`minutes`:`seconds`.`nanoseconds``offset`[`iana`]**

> if an *IANA-Timezone* is available.

Reviewer 4: What is the intended purpose of this condition? Don't ZonedDateTime objects always have an IANA time zone?

Author: Remnant of earlier versions

The `year` is 0-padded to a minimum of 4 digits. `month`, `day`, `hours`, `minutes`, `seconds`
are 0-padded to a minimum of 2 digits. `nanoseconds` is 0-padded to a minimum of 9 digits. The
`offset` is the timezone offset as created by `zoned.offset. `iana` is the *IANA-Timezone*
string.

Examples:

- `1976-11-18T15:23:30.123456789+01:00[Europe/Vienna]` - created with `Europe/Vienna` timezone
- `1976-11-18T15:23:30.123456789+01:00[Europe/Berlin]` - created with `Europe/Berlin` timezone

#### ZonedDateTime.prototype.toJSON() : string

Equivalent to `ZonedDateTime.prototype.toString() : string`

#### ZonedDateTime.fromString(iso: string) : ZonedDateTime

> Parses a `string` in the specific ISO-8601 format emitted by `ZonedDateTime.prototype.toString()`.

Reviewer 4: ```suggestion
Creates a new `ZonedDateTime` object from parsing a string that must be in the same ISO 8601 format used for `ZonedDateTime.prototype.toString()`, with precision at or exceeding the minute level, non-empty non-"Z" UTC offset representation, and a bracketed IANA time zone name in which the offset is correct for the represented date and time.
```

#### CivilDate.prototype.dayOfWeek : number

The `.dayOfWeek` property represents the day of the week where Monday is `1` and `Sunday` is `7` in accordance with ISO-8601.

#### CivilDate.prototype.dayOfYear : number

The `.dayOfYear` property represents the ordinal day of the Gregorian year according to ISO-8601.

#### CivilDate.prototype.weekOfYear : number

The `.weekOfYear` property represents the ISO week-number. Beware that dates at the begining of a year may be part of a week from the preceding year, and dates at the end of a year may be part of a week at the beginning of the next year, as the first week of any year is defined as the week that contains the first Thursday of the week.

#### CivilDate.prototype.plus(value: IntervalLike) : CivilDate

Creates a new `CivilDate` object by adding (subtracting for negative values) values to its members.
The specified values must be numeric if specified.

The algorithm is such that:
  1. the individual values are added to the existing values.
  2. the range of `days` is ensured to be between 1 and 29-31 depending on the month by adjusting `month`
  3. the range of `months` is ensured to be between 1 and 12 by adjusting the `years`.

> #### CivilDate.prototype.with(values: DateLike) : CivilDate

Reviewer 4: Are the timeline-anchored objects intentionally missing `with` methods?

Author: Nope. Oversight

Creates a new `CivilDate` object by overriding specified values to its members.
The specified values must be numeric if specified.

> #### CivilDate.prototype.withTime(time : CivilTime) : CivilDateTime

Reviewer 4: ```suggestion
#### CivilDate.prototype.withTime(time : TimeLike) : CivilDateTime
```

Author: This would be wrong, because it doesn't require `instanceof CivilTime` to be true

Reviewer 4: So you want to force `civilDate.withTime(new CivilTime(10, 23))`? Per #120, `civilDate.withTime({hour: 10, minute: 23})` (not to mention `civilDate.withTime(crossRealmCivilTime)`) should be accepted.

Combines this `CivilDate` with the passed `CivilTime` to create a new `CivilDateTime` object.

#### CivilDate.prototype.toString() : string

Equivalent to `date.toDateString()`

#### CivilDate.prototype.toJSON() : string

Equivalent to `date.toString()`

#### CivilDate.prototype.toDateTimeString() : string

`.toDateString()` creates an ISO-8601 compliant string in the format:

> **`year`-`month`-`day`**. The `year` is 0-padded to a minimum of 4 digits. `month` and `day`.

Reviewer 4: What is the purpose of a method on CivilDate whose name suggests that it depends upon or produces time data?

Author: It's an oversight and should be `toDateString()`
